src/core/server/elasticsearch/client/create_transport.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { IncomingHttpHeaders } from 'http';
+import type {
+  TransportRequestParams,
+  TransportRequestOptions,
+  TransportResult,
+} from '@elastic/transport';
+import type { TransportOptions } from '@elastic/transport/lib/Transport';
+import { Transport } from '@elastic/elasticsearch';
+import { isUnauthorizedError } from './errors';
+import { InternalUnauthorizedErrorHandler, isRetryResult } from './retry_unauthorized';
+
+type TransportClass = typeof Transport;
+
+export type ErrorHandlerAccessor = () => InternalUnauthorizedErrorHandler;
+
+const noop = () => undefined;
+
+export const createTransport = ({
+  getExecutionContext = noop,
+  getUnauthorizedErrorHandler,
+}: {
+  getExecutionContext?: () => string | undefined;
+  getUnauthorizedErrorHandler?: ErrorHandlerAccessor;
+}): TransportClass => {
+  class KibanaTransport extends Transport {
+    private headers: IncomingHttpHeaders = {};
+
+    constructor(options: TransportOptions) {
+      const { headers = {}, ...otherOptions } = options;
+      super(otherOptions);
+      this.headers = headers;
+    }
+
+    async request(params: TransportRequestParams, options?: TransportRequestOptions) {
+      const opts: TransportRequestOptions = { ...options } || {};
+      const optionHeaders = { ...opts.headers };
+      const opaqueId = getExecutionContext();
+      if (opaqueId && !opts.opaqueId) {
+        // rewrites headers['x-opaque-id'] if it presents
+        opts.opaqueId = opaqueId;
+      }
+      // Enforce the client to return TransportResult.
+      // It's required for bwc with responses in 7.x version.
+      if (opts.meta === undefined) {
+        opts.meta = true;
+      }
+
+      // add stored headers to the options
+      opts.headers = {
+        ...this.headers,
+        ...optionHeaders,
```

Comment:
nit: we copy `opts.headers` twice.
```suggestion
        ...opts.headers,
```

Reply:
I had incredible issues in unit tests because of the way the client keep references / mutates the headers that forced me to do that to assert against the values of the used headers for each individual call. but I will try again.